<b>1. Binary Classification with the Breast Cancer Dataset</b>

In [1]:
import numpy as np
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, classification_report

# Load dataset
data = load_breast_cancer()
X, y = data.data, data.target

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Scale features for better convergence
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

# Fit logistic regression model
clf = LogisticRegression(max_iter=10000, random_state=42)
clf.fit(X_train_scaled, y_train)

# Predictions and evaluation
y_pred = clf.predict(X_test_scaled)
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[ 62   1]
 [  2 106]]

Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.98      0.98        63
           1       0.99      0.98      0.99       108

    accuracy                           0.98       171
   macro avg       0.98      0.98      0.98       171
weighted avg       0.98      0.98      0.98       171



# Binary Classification Inferences

Based on the binary classification results from the breast cancer model, here are the key inferences:

1. **Overall High Accuracy:**
   - The model achieves an overall accuracy of **98%** on the test set, indicating that nearly all samples are correctly classified.

2. **Balanced Performance for Both Classes:**
   - **Class 0 (e.g., benign):**  
     - True negatives: 62  
     - False positives: 1  
   - **Class 1 (e.g., malignant):**  
     - True positives: 106  
     - False negatives: 2  
   This balance suggests that the model is highly effective in differentiating between the two classes.

3. **High Precision:**
   - **Class 0:** Precision = 0.97  
     (97% of the predictions for class 0 are correct)
   - **Class 1:** Precision = 0.99  
     (99% of the predictions for class 1 are correct)
   
   High precision minimizes the risk of misclassifying a benign case as malignant or vice versa.

4. **High Recall (Sensitivity):**
   - Both classes show a recall of **0.98**, meaning the model correctly identifies 98% of the actual instances in each class.
   - High recall for the malignant class is crucial in a clinical context to avoid missing true cancer cases.

5. **Strong F1-Scores:**
   - **Class 0:** F1-score = 0.98  
   - **Class 1:** F1-score = 0.99  
   These scores reflect an excellent balance between precision and recall.

6. **Clinical Implications:**
   - The model's robust performance (high accuracy, precision, recall, and F1-scores) suggests that it could serve as a reliable diagnostic tool.
   - Particularly, the high recall for malignant cases is vital for early detection, reducing the chance of missing true cancer cases.
   - The balanced performance across both classes helps build trust in its predictions, which is essential for clinical decision-making.




<b>2. Multinomial Classification with the Iris Dataset</b>

In [3]:
import warnings
warnings.filterwarnings('ignore')
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report

# Load iris dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Fit multinomial logistic regression model
clf_multi = LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=10000, random_state=42)
clf_multi.fit(X_train, y_train)

# Evaluate the model
y_pred_multi = clf_multi.predict(X_test)
print("Multinomial Logistic Regression")
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_multi))
print("\nClassification Report:")
print(classification_report(y_test, y_pred_multi))


Multinomial Logistic Regression
Confusion Matrix:
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        13

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45



# Multinomial Logistic Regression Inferences



1. **Perfect Classification Performance:**
   - The confusion matrix shows that each class is perfectly classified:
     - **Class 0:** 19 true predictions, 0 misclassifications.
     - **Class 1:** 13 true predictions, 0 misclassifications.
     - **Class 2:** 13 true predictions, 0 misclassifications.
   - The overall accuracy is **100%**, which means every sample in the test set was correctly classified.

2. **Model Metrics:**
   - **Precision, Recall, and F1-Score:** All are **1.00** for each class. This indicates that:
     - Every instance predicted to belong to a given class is correct (high precision).
     - All actual instances of each class are correctly identified (high recall).
     - The harmonic mean of precision and recall (F1-score) is perfect for every class.

3. **Dataset and Model Considerations:**
   - The results likely stem from a well-known, easily separable dataset (such as the Iris dataset), which is known to yield near-perfect results with many classification algorithms.
   - Although the performance is perfect on the given test set (45 samples), this might be due to the simplicity and small size of the dataset. For larger or more complex datasets, further evaluation and cross-validation would be essential to ensure the model's robustness.

4. **Practical Implications:**
   - The excellent performance suggests that multinomial logistic regression is highly effective for problems where classes are clearly distinct.
   - In practical applications, such high accuracy would instill confidence in the model's predictions, assuming the training and test data are representative of real-world conditions.

5. **Next Steps for Validation:**
   - Consider performing cross-validation or testing on a larger dataset to confirm that the model generalizes well beyond this specific test set.
   - Analyze feature importance and decision boundaries if further interpretability is required.




<b>3. Ordinal Classification with the Wine Quality Dataset</b>

In [5]:
import pandas as pd
import numpy as np
from statsmodels.miscmodels.ordinal_model import OrderedModel
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load the red wine quality dataset
wine = pd.read_csv("winequality-red.csv", sep=";")

# Define target (ordinal quality score) and predictors
y = wine["quality"]
X = wine.drop("quality", axis=1)

# Scale features for better convergence
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled = pd.DataFrame(X_scaled, columns=X.columns)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.3, random_state=42
)

# Fit an ordinal logistic regression (proportional odds) model using statsmodels
# Here we use the cumulative logit ("logit") model.
model = OrderedModel(y_train, X_train, distr="logit")
res = model.fit(method='bfgs', disp=False)
print(res.summary())

# Get predicted probabilities on the test set
predicted_probs = res.model.predict(res.params, exog=X_test)
# predicted_probs is a NumPy array with shape (n_samples, n_categories)

# Define the sorted list of quality levels from the training set
quality_levels = sorted(y_train.unique())
# For example, quality_levels might be [3, 4, 5, 6, 7, 8]

# Use np.argmax to find the index of the highest probability for each observation
predicted_index = np.argmax(predicted_probs, axis=1)

# Map these indices to the actual quality labels
predicted_class = [quality_levels[i] for i in predicted_index]

# Evaluate the model: Create a confusion matrix comparing actual vs. predicted quality levels
conf_matrix = pd.crosstab(y_test, predicted_class, rownames=["Actual"], colnames=["Predicted"])
print("\nOrdinal Logistic Regression Confusion Matrix:")
print(conf_matrix)


                             OrderedModel Results                             
Dep. Variable:                quality   Log-Likelihood:                -1072.0
Model:                   OrderedModel   AIC:                             2176.
Method:            Maximum Likelihood   BIC:                             2256.
Date:                Fri, 14 Mar 2025                                         
Time:                        08:30:36                                         
No. Observations:                1119                                         
Df Residuals:                    1103                                         
Df Model:                          11                                         
                           coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------
fixed acidity            0.2356      0.169      1.392      0.164      -0.096       0.567
volatile acidity      

# Ordinal Logistic Regression Inference

## Model Overview

- **Objective:**  
  The model predicts the ordinal outcome “quality” of wine using several chemical predictors, such as fixed acidity, volatile acidity, citric acid, residual sugar, chlorides, free sulfur dioxide, total sulfur dioxide, density, pH, sulphates, and alcohol.

- **Fit Statistics:**  
  - **Log-Likelihood:** -1072.0  
  - **AIC:** 2176  
  - **BIC:** 2256  
  These statistics provide insight into the model’s goodness-of-fit and can be used to compare with other models.

## Coefficient Interpretations

Each coefficient reflects the change in the log odds of a wine being in a higher quality category for a one-unit increase in the predictor, holding all else constant.

- **Volatile Acidity (coef = -0.6268, p < 0.001):**  
  A one-unit increase in volatile acidity decreases the odds of being in a higher quality category by a factor of approximately exp(-0.6268) ≈ 0.534. This is statistically significant.

- **Alcohol (coef = 0.9371, p < 0.001):**  
  Each one-unit increase in alcohol content increases the odds of a higher quality rating by a factor of exp(0.9371) ≈ 2.553, indicating a strong positive effect.

- **Sulphates (coef = 0.4351, p < 0.001):**  
  Higher sulphate levels are significantly associated with increased wine quality, with an odds ratio of exp(0.4351) ≈ 1.545.

- **Total Sulfur Dioxide (coef = -0.3651, p < 0.001):**  
  Increasing total sulfur dioxide by one unit reduces the odds of a higher quality rating by exp(-0.3651) ≈ 0.694.

- **Other Variables:**  
  - *Fixed Acidity (0.2356, p = 0.164), Citric Acid (-0.2011, p = 0.062), Residual Sugar (0.0947, p = 0.251), Density (-0.1410, p = 0.358), and pH (-0.0697, p = 0.526)* did not achieve conventional significance (p < 0.05).  
  - Citric acid is marginal (p ≈ 0.062) and might warrant further investigation with additional data or refined modeling.

## Thresholds (Cutpoints)

The model estimates several thresholds (3/4, 4/5, 5/6, 6/7, 7/8) which serve as boundaries on the latent variable separating the ordinal categories. These cutpoints help in distinguishing between the different quality ratings.

## Confusion Matrix

The confusion matrix compares the predicted quality classes (5, 6, and 7) with the actual quality classes (ranging from 3 to 8):

| Actual \ Predicted | 5   | 6   | 7   |
|--------------------|-----|-----|-----|
| **3**              | 1   | 0   | 0   |
| **4**              | 12  | 5   | 0   |
| **5**              | 147 | 48  | 0   |
| **6**              | 72  | 116 | 12  |
| **7**              | 2   | 47  | 12  |
| **8**              | 0   | 2   | 4   |

- **Interpretation:**  
  The majority of predictions are concentrated in the middle categories (5, 6, and 7). However, the model shows some misclassifications, particularly at the extreme ends (actual classes 3, 4, and 8). This suggests that while the model performs well for moderate quality wines, its predictive power diminishes for wines at the very low or high quality extremes.

## Overall Inference

- **Key Predictors:**  
  Significant determinants of wine quality include:
  - **Volatile Acidity:** Negative impact (higher acidity decreases quality).
  - **Alcohol:** Positive impact (higher alcohol content increases quality).
  - **Sulphates:** Positive impact (increases likelihood of higher quality).
  - **Total Sulfur Dioxide:** Negative impact (higher levels decrease quality).

- **Non-significant Variables:**  
  Fixed acidity, residual sugar, density, and pH did not show statistically significant effects. Citric acid’s borderline significance suggests it may have a minor influence that requires further examination.

- **Model Performance:**  
  The model fits the moderate quality ratings (5, 6, 7) better than the extreme ratings (3, 4, 8). This indicates that while the selected predictors are effective for the central range of wine quality, additional predictors or model adjustments might be needed to accurately capture the extremes.

In summary, the ordinal logistic regression analysis indicates that chemical properties such as volatile acidity, alcohol, sulphates, and total sulfur dioxide are key determinants of wine quality. The model effectively discriminates between middle quality levels but may benefit from refinement to improve predictions at the very low and high ends of the quality spectrum.
